target/linux/ar71xx/files/arch/mips/ath79/mach-tl-wr942n-v1.c

```diff
@@ -0,0 +1,252 @@
+#include "dev-spi.h"
+#include "dev-usb.h"
+#include "dev-wmac.h"
+#include "nvram.h"
+
+#define TL_WR942N_V1_KEYS_POLL_INTERVAL	20
+#define TL_WR942N_V1_KEYS_DEBOUNCE_INTERVAL	(3 * TL_WR942N_V1_KEYS_POLL_INTERVAL)
+
+#define TL_WR942N_V1_GPIO_BTN_RESET		1
+#define TL_WR942N_V1_GPIO_BTN_RFKILL		2
+
+#define TL_WR942N_V1_GPIO_LED_STATUS		22
+#define TL_WR942N_V1_GPIO_LED_WPS		21
+#define TL_WR942N_V1_GPIO_SHIFT_OE		16
+#define TL_WR942N_V1_GPIO_SHIFT_SER		17
+#define TL_WR942N_V1_GPIO_SHIFT_SRCLK		18
+#define TL_WR942N_V1_GPIO_SHIFT_SRCLR		19
+#define TL_WR942N_V1_GPIO_SHIFT_RCLK		20
+
+#define TL_WR942N_V1_74HC_GPIO_BASE		QCA956X_GPIO_COUNT
+#define TL_WR942N_V1_74HC_GPIO_LED_LAN4		23
+#define TL_WR942N_V1_74HC_GPIO_LED_LAN3		24
+#define TL_WR942N_V1_74HC_GPIO_LED_LAN2		25
+#define TL_WR942N_V1_74HC_GPIO_LED_LAN1		26
+#define TL_WR942N_V1_74HC_GPIO_LED_WAN_GREEN	27
+#define TL_WR942N_V1_74HC_GPIO_LED_WAN_AMBER	28
+#define TL_WR942N_V1_74HC_GPIO_LED_WLAN		29
+
+#define TL_WR942N_V1_SSR_BIT_0			0
+#define TL_WR942N_V1_SSR_BIT_1			1
+#define TL_WR942N_V1_SSR_BIT_2			2
+#define TL_WR942N_V1_SSR_BIT_3			3
+#define TL_WR942N_V1_SSR_BIT_4			4
+#define TL_WR942N_V1_SSR_BIT_5			5
+#define TL_WR942N_V1_SSR_BIT_6			6
+#define TL_WR942N_V1_SSR_BIT_7			7
+
+#define TL_WR942N_V1_WMAC_CALDATA_OFFSET	0x1000
+#define TL_WR942N_V1_DEFAULT_MAC_ADDR		0x1fe40008
+#define TL_WR942N_V1_DEFAULT_MAC_SIZE		0x200
+
+static struct gpio_led tl_wr942n_v1_leds_gpio[] __initdata = {
+	{
+		.name		= "tl-wr942n-v1:green:status",
+		.gpio		= TL_WR942N_V1_GPIO_LED_STATUS,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:green:wlan",
+		.gpio		= TL_WR942N_V1_74HC_GPIO_LED_WLAN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:green:lan1",
+		.gpio		= TL_WR942N_V1_74HC_GPIO_LED_LAN1,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:green:lan2",
+		.gpio		= TL_WR942N_V1_74HC_GPIO_LED_LAN2,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:green:lan3",
+		.gpio		= TL_WR942N_V1_74HC_GPIO_LED_LAN3,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:green:lan4",
+		.gpio		= TL_WR942N_V1_74HC_GPIO_LED_LAN4,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:green:wan",
+		.gpio		= TL_WR942N_V1_74HC_GPIO_LED_WAN_GREEN,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:amber:wan",
+		.gpio		= TL_WR942N_V1_74HC_GPIO_LED_WAN_AMBER,
+		.active_low	= 1,
+	},
+	{
+		.name		= "tl-wr942n-v1:green:wps",
+		.gpio		= TL_WR942N_V1_GPIO_LED_WPS,
+		.active_low	= 1,
+	},
+};
+
+static struct gpio_keys_button tl_wr942n_v1_gpio_keys[] __initdata = {
+	{
+		.desc			= "Reset button",
+		.type			= EV_KEY,
+		.code			= KEY_RESTART,
+		.debounce_interval	= TL_WR942N_V1_KEYS_DEBOUNCE_INTERVAL,
+		.gpio			= TL_WR942N_V1_GPIO_BTN_RESET,
+		.active_low		= 1,
+	},
+	{
+		.desc			= "RFKILL button",
+		.type			= EV_KEY,
+		.code			= KEY_RFKILL,
+		.debounce_interval	= TL_WR942N_V1_KEYS_DEBOUNCE_INTERVAL,
+		.gpio			= TL_WR942N_V1_GPIO_BTN_RFKILL,
+		.active_low		= 1,
+	},
+/*	{
```

Comment:
Please, drop code which is commented out anyway or include information why it should be here.